lib/vector-core/src/config/mod.rs

```diff
@@ -100,42 +100,129 @@ impl Input {
+    ///
+    ///
+    /// Sets the datatype to be [`DataType::Trace`].
+    #[must_use]
+    pub fn new_traces() -> Self {
+        Self {
+            port: None,
+            ty: DataType::Trace,
+            schema_definition: None,
         }
     }
 
-    /// Set the schema definition for this `Output`.
+    /// Return the schema [`schema::Definition`] from this output.
+    ///
+    /// Takes a `schema_enabled` flag to determine if the full definition including the fields
+    /// and associated types should be returned, or if a simple definition should be returned.
+    /// Schema enabled is set in the users configuration.
     #[must_use]
-    pub fn with_schema_definition(mut self, schema_definition: schema::Definition) -> Self {
-        self.log_schema_definition = Some(schema_definition);
+    pub fn schema_definition(&self, schema_enabled: bool) -> Option<schema::Definition> {
+        self.schema_definition.as_ref().map(|definition| {
+            if schema_enabled || definition.log_namespaces().contains(&LogNamespace::Vector) {
+                definition.clone()
+            } else {
+                let mut new_definition =
+                    schema::Definition::default_for_namespace(definition.log_namespaces());
+                new_definition.add_meanings(definition.meanings());
+                new_definition
+            }
+        })
+    }
+
+    /// Consumes the object and returns the schema [`schema::Definition`] from this output.
+    ///
+    /// Takes a `schema_enabled` flag to determine if the full definition including the fields
+    /// and associated types should be returned, or if a simple definition should be returned.
+    /// Schema enabled is set in the user's configuration.
+    #[must_use]
+    pub fn into_schema_definition(self, schema_enabled: bool) -> Option<schema::Definition> {
+        self.schema_definition.map(|definition| {
+            if schema_enabled || definition.log_namespaces().contains(&LogNamespace::Vector) {
+                definition
+            } else {
+                let mut new_definition =
+                    schema::Definition::default_for_namespace(definition.log_namespaces());
+                new_definition.add_meanings(definition.meanings());
+                new_definition
+            }
+        })
+    }
+}
+
+impl SourceOutput {
+    /// Set the port name for this `SourceOutput`.
+    #[must_use]
+    pub fn with_port(mut self, name: impl Into<String>) -> Self {
+        self.port = Some(name.into());
         self
     }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TransformOutput {
+    pub port: Option<String>,
+    pub ty: DataType,
+
+    /// For *transforms* if `Datatype` is [`DataType::Log`], at least one definition should be output. If the
+    /// transform has multiple connected sources, it is possible to have multiple output definitions -
+    /// one for each input.
+    pub log_schema_definitions: Vec<schema::Definition>,
```

Comment:
We are still planning on making this a `HashMap<OutputId, schema::Definition>` in the future, right?

Reply:
Yes. That's the next PR.